include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h

```diff
@@ -252,7 +306,66 @@ class MultiVoxelGridCovariance : public pcl::VoxelGrid<PointT> {
   // Return the string indices of currently loaded map pieces
   std::vector<std::string> getCurrentMapIDs() const;
 
+  void setThreadNum(int thread_num) {
+    sync();
+
+    if(thread_num <= 0) {
+      thread_num_ = 1;
+    }
+
+    thread_num_ = thread_num;
+    thread_futs_.resize(thread_num_);
+    processing_inputs_.resize(thread_num_);
+  }
+
+  ~MultiVoxelGridCovariance() {
+    // Stop all threads in the case an intermediate interrupt occurs
+    sync();
+  }
+
 protected:
+  // Return the index of an idle thread, which is not running any
+  // job, or has already finished its job and waiting for a join.
+  inline int get_idle_tid() {
+    int tid = (last_check_tid_ == thread_num_ - 1) ? 0 : last_check_tid_ + 1;
+    std::chrono::microseconds span(50);
+
+    // Loop until an idle thread is found
+    while(true) {
+      // Return immediately if a thread that has not been given a job is found
+      if(!thread_futs_[tid].valid()) {
+        last_check_tid_ = tid;
+        return tid;
+      }
+
+      // If no such thread is found, wait for the current thread to finish its job
+      if(thread_futs_[tid].wait_for(span) == std::future_status::ready) {
+        last_check_tid_ = tid;
+        return tid;
+      }
+
+      // If the current thread has not finished its job, check the next thread
+      tid = (tid == thread_num_ - 1) ? 0 : tid + 1;
+    }
+  }
+
+  // Wait for all running threads to finish
+  inline void sync() {
+    for(int i = 0; i < thread_futs_.size(); ++i) {
```

Comment:
Sorry, additional comment.
autoware.universe strictly checks warnings, so when I built `ndt_scan_matcher`, the following error occurred.

```
+ colcon build --symlink-install --cmake-args -DCMAKE_BUILD_TYPE=Release --packages-up-to ndt_scan_matcher
Starting >>> ndt_scan_matcher
[Processing: ndt_scan_matcher]                               
--- stderr: ndt_scan_matcher                                     
In file included from /home/shintarosakoda/autoware/install/ndt_omp/include/multigrid_pclomp/multigrid_ndt_omp.h:62,
                 from /home/shintarosakoda/autoware/src/universe/autoware.universe/localization/ndt_scan_matcher/include/ndt_scan_matcher/hyper_parameters.hpp:20,
                 from /home/shintarosakoda/autoware/src/universe/autoware.universe/localization/ndt_scan_matcher/include/ndt_scan_matcher/map_update_module.hpp:19,
                 from /home/shintarosakoda/autoware/src/universe/autoware.universe/localization/ndt_scan_matcher/src/map_update_module.cpp:15:
/home/shintarosakoda/autoware/install/ndt_omp/include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h: In instantiation of 'void pclomp::MultiVoxelGridCovariance<PointT>::sync() [with PointT = pcl::PointXYZ]':
/home/shintarosakoda/autoware/install/ndt_omp/include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h:310:5:   required from 'void pclomp::MultiVoxelGridCovariance<PointT>::setThreadNum(int) [with PointT = pcl::PointXYZ]'
/home/shintarosakoda/autoware/install/ndt_omp/include/multigrid_pclomp/multigrid_ndt_omp.h:321:31:   required from 'void pclomp::MultiGridNormalDistributionsTransform<PointSource, PointTarget>::setParams(const pclomp::NdtParams&) [with PointSource = pcl::PointXYZ; PointTarget = pcl::PointXYZ]'
/home/shintarosakoda/autoware/src/universe/autoware.universe/localization/ndt_scan_matcher/src/map_update_module.cpp:79:24:   required from here
/home/shintarosakoda/autoware/install/ndt_omp/include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h:354:22: error: comparison of integer expressions of different signedness: 'int' and 'std::vector<std::future<bool> >::size_type' {aka 'long unsigned int'} [-Werror=sign-compare]
  354 |     for(int i = 0; i < thread_futs_.size(); ++i) {
      |                    ~~^~~~~~~~~~~~~~~~~~~~~
In file included from /home/shintarosakoda/autoware/install/ndt_omp/include/multigrid_pclomp/multigrid_ndt_omp.h:62,
                 from /home/shintarosakoda/autoware/src/universe/autoware.universe/localization/ndt_scan_matcher/include/ndt_scan_matcher/hyper_parameters.hpp:20,
                 from /home/shintarosakoda/autoware/src/universe/autoware.universe/localization/ndt_scan_matcher/include/ndt_scan_matcher/map_update_module.hpp:19,
                 from /home/shintarosakoda/autoware/src/universe/autoware.universe/localization/ndt_scan_matcher/src/map_update_module.cpp:15:
/home/shintarosakoda/autoware/install/ndt_omp/include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h: In instantiation of 'void pclomp::MultiVoxelGridCovariance<PointT>::sync() [with PointT = pcl::PointXYZ]':
/home/shintarosakoda/autoware/install/ndt_omp/include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h:310:5:   required from 'void pclomp::MultiVoxelGridCovariance<PointT>::setThreadNum(int) [with PointT = pcl::PointXYZ]'
/home/shintarosakoda/autoware/install/ndt_omp/include/multigrid_pclomp/multigrid_ndt_omp.h:321:31:   required from 'void pclomp::MultiGridNormalDistributionsTransform<PointSource, PointTarget>::setParams(const pclomp::NdtParams&) [with PointSource = pcl::PointXYZ; PointTarget = pcl::PointXYZ]'
/home/shintarosakoda/autoware/src/universe/autoware.universe/localization/ndt_scan_matcher/src/map_update_module.cpp:79:24:   required from here
/home/shintarosakoda/autoware/install/ndt_omp/include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h:354:22: error: comparison of integer expressions of different signedness: 'int' and 'std::vector<std::future<bool> >::size_type' {aka 'long unsigned int'} [-Werror=sign-compare]
  354 |     for(int i = 0; i < thread_futs_.size(); ++i) {
      |                    ~~^~~~~~~~~~~~~~~~~~~~~
In file included from /home/shintarosakoda/autoware/install/ndt_omp/include/multigrid_pclomp/multigrid_ndt_omp.h:62,
                 from /home/shintarosakoda/autoware/src/universe/autoware.universe/localization/ndt_scan_matcher/include/ndt_scan_matcher/hyper_parameters.hpp:20,
                 from /home/shintarosakoda/autoware/src/universe/autoware.universe/localization/ndt_scan_matcher/include/ndt_scan_matcher/ndt_scan_matcher_core.hpp:21,
                 from /home/shintarosakoda/autoware/src/universe/autoware.universe/localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp:15:
/home/shintarosakoda/autoware/install/ndt_omp/include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h: In instantiation of 'void pclomp::MultiVoxelGridCovariance<PointT>::sync() [with PointT = pcl::PointXYZ]':
/home/shintarosakoda/autoware/install/ndt_omp/include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h:310:5:   required from 'void pclomp::MultiVoxelGridCovariance<PointT>::setThreadNum(int) [with PointT = pcl::PointXYZ]'
/home/shintarosakoda/autoware/install/ndt_omp/include/multigrid_pclomp/multigrid_ndt_omp.h:321:31:   required from 'void pclomp::MultiGridNormalDistributionsTransform<PointSource, PointTarget>::setParams(const pclomp::NdtParams&) [with PointSource = pcl::PointXYZ; PointTarget = pcl::PointXYZ]'
/home/shintarosakoda/autoware/src/universe/autoware.universe/localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp:178:22:   required from here
/home/shintarosakoda/autoware/install/ndt_omp/include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h:354:22: error: comparison of integer expressions of different signedness: 'int' and 'std::vector<std::future<bool> >::size_type' {aka 'long unsigned int'} [-Werror=sign-compare]
  354 |     for(int i = 0; i < thread_futs_.size(); ++i) {
      |                    ~~^~~~~~~~~~~~~~~~~~~~~
In file included from /home/shintarosakoda/autoware/install/ndt_omp/include/multigrid_pclomp/multigrid_ndt_omp.h:62,
                 from /home/shintarosakoda/autoware/src/universe/autoware.universe/localization/ndt_scan_matcher/include/ndt_scan_matcher/hyper_parameters.hpp:20,
                 from /home/shintarosakoda/autoware/src/universe/autoware.universe/localization/ndt_scan_matcher/include/ndt_scan_matcher/ndt_scan_matcher_core.hpp:21,
                 from /home/shintarosakoda/autoware/src/universe/autoware.universe/localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp:15:
/home/shintarosakoda/autoware/install/ndt_omp/include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h: In instantiation of 'void pclomp::MultiVoxelGridCovariance<PointT>::sync() [with PointT = pcl::PointXYZ]':
/home/shintarosakoda/autoware/install/ndt_omp/include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h:310:5:   required from 'void pclomp::MultiVoxelGridCovariance<PointT>::setThreadNum(int) [with PointT = pcl::PointXYZ]'
/home/shintarosakoda/autoware/install/ndt_omp/include/multigrid_pclomp/multigrid_ndt_omp.h:321:31:   required from 'void pclomp::MultiGridNormalDistributionsTransform<PointSource, PointTarget>::setParams(const pclomp::NdtParams&) [with PointSource = pcl::PointXYZ; PointTarget = pcl::PointXYZ]'
/home/shintarosakoda/autoware/src/universe/autoware.universe/localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp:178:22:   required from here
/home/shintarosakoda/autoware/install/ndt_omp/include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h:354:22: error: comparison of integer expressions of different signedness: 'int' and 'std::vector<std::future<bool> >::size_type' {aka 'long unsigned int'} [-Werror=sign-compare]
  354 |     for(int i = 0; i < thread_futs_.size(); ++i) {
      |                    ~~^~~~~~~~~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors
gmake[2]: *** [CMakeFiles/once_initialize_at_out_of_map_then_initialize_correctly.dir/build.make:118: CMakeFiles/once_initialize_at_out_of_map_then_initialize_correctly.dir/src/map_update_module.cpp.o] エラー 1
gmake[2]: *** 未完了のジョブを待っています....
cc1plus: all warnings being treated as errors
gmake[2]: *** [CMakeFiles/standard_sequence_for_initial_pose_estimation.dir/build.make:118: CMakeFiles/standard_sequence_for_initial_pose_estimation.dir/src/map_update_module.cpp.o] エラー 1
gmake[2]: *** 未完了のジョブを待っています....
cc1plus: all warnings being treated as errors
gmake[2]: *** [CMakeFiles/once_initialize_at_out_of_map_then_initialize_correctly.dir/build.make:104: CMakeFiles/once_initialize_at_out_of_map_then_initialize_correctly.dir/src/ndt_scan_matcher_core.cpp.o] エラー 1
gmake[1]: *** [CMakeFiles/Makefile2:213: CMakeFiles/once_initialize_at_out_of_map_then_initialize_correctly.dir/all] エラー 2
gmake[1]: *** 未完了のジョブを待っています....
cc1plus: all warnings being treated as errors
gmake[2]: *** [CMakeFiles/standard_sequence_for_initial_pose_estimation.dir/build.make:104: CMakeFiles/standard_sequence_for_initial_pose_estimation.dir/src/ndt_scan_matcher_core.cpp.o] エラー 1
gmake[1]: *** [CMakeFiles/Makefile2:186: CMakeFiles/standard_sequence_for_initial_pose_estimation.dir/all] エラー 2
gmake: *** [Makefile:146: all] エラー 2
---
Failed   <<< ndt_scan_matcher [42.0s, exited with code 2]

Summary: 34 packages finished [1min 2s]
  1 package failed: ndt_scan_matcher
  1 package had stderr output: ndt_scan_matcher
```

Reply:
@SakodaShintaro I fixed this one but do you know why this error did not show up when I build the code on my local PC?

Reply:
Hmm, the only thing I can think of is that the build cache remained. Thank you for the fixing.